netket/vqs/mc/mc_state/state.py

```diff
@@ -707,9 +717,10 @@ def local_estimators(
 def serialize_MCState(vstate):
     state_dict = {
         "variables": serialization.to_state_dict(vstate.variables),
-        "sampler_state": serialization.to_state_dict(vstate.sampler_state),
+        "sampler_state": serialization.to_state_dict(vstate._sampler_state_previous),
```

Comment:
what is sampler state_previous? isn't this a bit too specific for markov chain?

Reply:
no it's a general problem of all samplers: the situation is the following:

```python
vs = #state with no samples, rng_state = 1
vs.samples # returns samples(rng_state=1)
# vs now has rng_state=2, samples(rng_state=1)

#now we serialise vstate
binary_data = serialise(vs) # we store here rng_state = 2, we don't store the samples
vs_new = deserialise(vs) # here we have rng_state=2, no samples
vs_new.samples # returns samples(rng_state=2)
```

so to make serialization/deserialization work properly with the caching of the samples, we need to store the old rng key (`sampler_state` in practice) *before* we had sampled.